remerkleable/basic.py

```diff
@@ -73,21 +76,98 @@ def __new__(cls, value: int):
 
-    def __sub__(self, other):
+    def __radd__(self: T, other: int) -> T:
+        return self.__add__(other)
+
+    def __sub__(self: T, other: int) -> T:
         return self.__class__(super().__sub__(self.__class__.coerce_view(other)))
 
-    def __mul__(self, other):
+    def __rsub__(self: T, other: int) -> T:
+        return self.__class__(self.__class__.coerce_view(other).__sub__(self))
+
+    def __mul__(self: T, other: int) -> T:
         return self.__class__(super().__mul__(self.__class__.coerce_view(other)))
 
-    def __floordiv__(self, other):  # Better known as "//"
+    def __rmul__(self: T, other: int) -> T:
+        return self.__mul__(other)
+
+    def __mod__(self: T, other: int) -> T:
+        return self.__class__(super().__mod__(self.__class__.coerce_view(other)))
+
+    def __rmod__(self: T, other: int) -> T:
+        return self.__class__(self.__class__.coerce_view(other).__mod__(self))
+
+    def __floordiv__(self: T, other: int) -> T:  # Better known as "//"
         return self.__class__(super().__floordiv__(self.__class__.coerce_view(other)))
 
-    def __truediv__(self, other):
+    def __rfloordiv__(self: T, other: int) -> T:
+        return self.__class__(self.__class__.coerce_view(other).__floordiv__(self))
+
+    def __truediv__(self: T, other: int) -> T:
         raise OperationNotSupported(f"non-integer division '{self} / {other}' "
-                                    f"is not valid for {self.__class__.type_repr()} type")
+                                    f"is not valid for {self.__class__.type_repr()} left hand type")
+
+    def __rtruediv__(self: T, other: int) -> T:
+        raise OperationNotSupported(f"non-integer division '{other} / {self}' "
+                                    f"is not valid for {self.__class__.type_repr()} right hand type")
+
+    def __pow__(self: T, other: int, modulo=None) -> T:
+        return self.__class__(super().__pow__(other, modulo))  # TODO: stricter argument checks?
+
+    def __rpow__(self: T, other, modulo=None) -> T:
+        return self.__class__(super().__rpow__(other, modulo))  # TODO: see __pow__
+
+    def __lshift__(self: T, other: int) -> T:
+        """Left bitshift clips bits at uint boundary"""
+        mask = (1 << (self.type_byte_length() << 3)) - 1
+        return self.__class__(super().__lshift__(int(other)) & mask)
+
+    def __rlshift__(self: T, other: int) -> T:
+        raise OperationNotSupported(f"{other} << {self} through __rlshift__ is not supported, "
+                                    f"{other} must be a uint type with __lshift__")
+
+    def __rshift__(self: T, other: int) -> T:
+        return self.__class__(super().__rshift__(int(other)))
+
+    def __rrshift__(self: T, other: int) -> T:
+        raise OperationNotSupported(f"{other} >> {self} through __rrshift__ is not supported, "
+                                    f"{other} must be a uint type with __rshift__")
```

Comment:
@protolambda 
There is a `_rrshift` case in [`compute_shuffled_index`](https://github.com/ethereum/eth2.0-specs/blob/hwwhww/strict-uint64/specs/phase0/beacon-chain.md#compute_shuffled_index)

```python
bit = (byte >> (position % 8)) % 2
```
where `position` is `uint64` type.

Do you think we should support `__rrshift__`? A quick workaround is casting `int` as below, but it's suboptimal for this PR's goal.
```python
bit = (byte >> int(position % 8)) % 2
```

Reply:
Oh, not having `>>` between two uints was not the intention. The worry is that `a << b` has an undefined output when `a` has no size. And similarly I think `>>` should not do anything when `a` is not a `uint`. We can't tell from the uint size of the shift, as e.g. `uint8(123) >> uint64(3)` may be a thing. I'll change it to raise a value-error when `other` is not a `uint`.

Reply:
Fixed here: https://github.com/protolambda/remerkleable/pull/7